### ISSUE TITLE
macOS cannot verify the developer of "KeePassXC".

### ISSUE THREAD

**Issue by gxfgbo** [keepassxreboot/keepassxc#7630]
Unable to install KeePassXC 2.7.0 on macOS 12.3 without accepting install as an unidentified developer.
Running Intel Core i7

**Steps to Reproduce**
      Download, verify sha and mount KeePassXC-2.7.0-x86_64.dmg
      Move .app to Application folder

**Expected Behavior**
      App is successfully installed and launchable without warning about developer's identity

**Actual Behavior**
      Prompted with: macOS cannot verify the developer of "KeePassXC". Are you sure you want to open it?

**Context**
      Occurred pending a macOS update. Rebooted and updated to 12.3. Redownloaded .dmg and issue still occurs. 
      Not possible to paste "Help -> About here" as KeePassXC is not launching

Operating System: macOS 12.3 
Desktop Env: N/A
Windowing System: N/A

**Comment by phoerious** [comment]
Are you sure you have the correct DMG? The DMG is signed with my developer ID. Please post a screenshot of the message you are seeing.

**Comment by varjolintu** [comment]
Make sure you are downloading the DMG directly from our site.

**Comment by gxfgbo** [comment]
I downloaded the installer from https://github.com/keepassxreboot/keepassxc/releases/download/2.7.0/KeePassXC-2.7.0-x86_64.dmg

I downloaded the SHA digest from https://github.com/keepassxreboot/keepassxc/releases/download/2.7.0/KeePassXC-2.7.0-x86_64.dmg.DIGEST

I run shasum in the same directory as the downloaded files: shasum -a 256 -c KeePassXC-*.DIGEST
KeePassXC-2.7.0-x86_64.dmg: OK

FYI: shasum -a 256 KeePassXC-2.7.0-x86_64.dmg 
777f03f6bc5bf7fdb45cf72017abb59883933cebdfac12febaf9fe3197dd1195  KeePassXC-2.7.0-x86_64.dmg

I receive the following error after double-clicking on the .dmg and trying to drag the KeePassXC app into the Applications folder:
<img width="259" alt="image" src="https://user-images.githubusercontent.com/60147752/159910894-e252ef57-28bd-4a41-800a-a53e5da3d630.png">

Let me know if it would be helpful to upload my CodeResources file somewhere and how to do so.

FYI: shasum -a 256 CodeResources        
914ab78ce88f3f2a95bbab6f1e651897c3f7f5caf7173f38f13afd352d1d9c04  CodeResources

Here is my system info:
<img width="245" alt="image" src="https://user-images.githubusercontent.com/60147752/159911785-46674371-6530-46e7-b655-1221ceec433f.png">

I should add, I experienced the same issue immediately prior to updating my Mac from 12.2.1 to 12.3.

**Comment by varjolintu** [comment]
What does `codesign -dv --verbose=4 /Applications/KeePassXC.app` show?

**Comment by phoerious** [comment]
Okay, that hints at a codesigning or rather a signature verification issue. I haven't seen that dialog in a while.

> Let me know if it would be helpful to upload my CodeResources file somewhere and how to do so.

If anything, you should compare the DMG digest to the SHA256 digests we publish on our website.

**Comment by gxfgbo** [comment]
> What does `codesign -dv --verbose=4 /Applications/KeePassXC.app` show?

I didn't allow the app to install which, I believe, would prevent me from being able to do as you suggest. Please let me know if I've misunderstood.

I can allow the app to install by bypassing the signature verification through System Preferences > Security and try checking the codesign if that would help.




> Okay, that hints at a codesigning or rather a signature verification issue. I haven't seen that dialog in a while.

I agree. I'm not familiar with how to debug signature verification other than ensuring the DMG is uncorrupted, ensuring SHAs match and @varjolintu's suggestion above.




> If anything, you should compare the DMG digest to the SHA256 digests we publish on our website.

Sorry I wasn't explicit. I did do this upon downloading the DMG and the DIGEST from the website.

FYI:
shasum -a 256 KeePassXC-2.7.0-x86_64.dmg
777f03f6bc5bf7fdb45cf72017abb59883933cebdfac12febaf9fe3197dd1195 KeePassXC-2.7.0-x86_64.dmg

more KeePassXC-2.7.0-x86_64.dmg.DIGEST  
777f03f6bc5bf7fdb45cf72017abb59883933cebdfac12febaf9fe3197dd1195  KeePassXC-2.7.0-x86_64.dmg

**Comment by droidmonkey** [comment]
I think this is just macOS being hyper paranoid about a file downloaded from a browser. You can obviously ignore the message because that is our official release package. Mac constantly subverts deployments outside of the app store.

**Comment by phoerious** [comment]
No, this is definitely some sort of signature verification issue. Otherwise you would have a button to open the application, not just to move it to trash. It could be an issue with expired root certificates or something else that is not on our end. At least I haven't seen anyone else having issues with this (including myself).

The sentence "cannot verify that this app is free from malware" may also point to a notarization verification issue, which is strange, because the ticket is also stapled onto the DMG in case you have no internet connection.

**Comment by tkoepsell** [comment]
I've had a similar problem, but with a different message: "KeePassXC is damaged and can't be opened.  You should move it to the Trash."  This has happened repeatedly after downloading the dmg-file and DIGEST from the KeePassXC website via Brave, Firefox, or Safari.  Running shasum always produces "KeePassXC-2.7.0-x86_64.dmg: OK".

The error message has a "?" button that leads to a MacOS Help page, which states:

   "The app has been modified, and its code does not match the original signed code. The app may be broken or corrupted, or it may have been tampered with. If you think that an app has been damaged or tampered with, you should delete it by dragging it to the Trash.  Some apps and tools, such as AppleScript or JavaScript applications and some legacy tools, modify themselves after signing. These types of apps cannot be opened unless you override the security settings on your Mac. See Open an app by overriding security settings."

If I try to ignore the error message and attempt to open KeePassXC anyway from the Applications folder using "Open," and then accept that it's from an unidentified developer, the KeePassXC icon appears in the dock, but the program hangs immediately.  The above error message continues to appear.  Pressing either "Move it to the trash" or "Cancel" quits the hanged application.

I've been unable to find a work-around.  I'm running MacOS 12.3 on an iMac desktop.

(Incidentally, the download page says that the current version for Intel processors is for MacOS 15+, which does not exist to my knowledge.  The current MacOS version is 12.3.)

**Comment by gxfgbo** [comment]
> I've had a similar problem, but with a different message: "KeePassXC is damaged and can't be opened. You should move it to the Trash."

I experience the same issue if I try to copy KeePassXC into the Applications folder twice. After the first copy I get the **"KeePassXC" cannot be opened because the developer cannot be verified.** message. Without acting on that message, copying KeePassXC into the Applications folder a second time yields the **"KeePassXC" is damaged and can't be opened.** message.

Anytime the **"KeePassXC" cannot be opened because the developer cannot be verified.** message is visible, the application resides in the Applications folder. I was able to run a codesign on it  

/Applications % codesign -dv --verbose=4 KeePassXC.app
Executable=/Applications/KeePassXC.app/Contents/MacOS/KeePassXC
Identifier=org.keepassxc.keepassxc
Format=app bundle with Mach-O thin (x86_64)
CodeDirectory v=20500 size=43651 flags=0x10000(runtime) hashes=1353+7 location=embedded
VersionPlatform=1
VersionMin=658688
VersionSDK=721664
Hash type=sha256 size=32
CandidateCDHash sha256=07b45df21d5882462ec0e192c2fa42b8dcf76244
CandidateCDHashFull sha256=07b45df21d5882462ec0e192c2fa42b8dcf762441d5c69063f490d135aab1dca
Hash choices=sha256
CMSDigest=07b45df21d5882462ec0e192c2fa42b8dcf762441d5c69063f490d135aab1dca
CMSDigestType=2
Executable Segment base=0
Executable Segment limit=5013504
Executable Segment flags=0x1
Page size=4096
CDHash=07b45df21d5882462ec0e192c2fa42b8dcf76244
Signature size=8931
Authority=Developer ID Application: Janek Bevendorff (G2S7P7J672)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=Mar 21, 2022 at 8:14:40 PM
Info.plist entries=15
TeamIdentifier=G2S7P7J672
Runtime Version=11.3.0
Sealed Resources version=2 rules=13 files=117
Internal requirements count=1 size=184

**Comment by gxfgbo** [comment]
:)

TLDR; Disable browser integration or exit all browsers before doing an upgrade.

I searched the mac console and found this:
ASP: Security policy would not allow process: 6774, /Applications/KeePassXC.app/Contents/MacOS/keepassxc-proxy

I closed and exited the browser and tried reinstalling. No developer verification error messages.

Upon starting either the browser or KeePassXC app after installing, I am now prompted with the following expected message:

<img width="255" alt="image" src="https://user-images.githubusercontent.com/60147752/159979595-511fd453-6d78-4cac-acc4-ce44717d3753.png">

I also tested disabling browser integration within the KeePassXC app before installing a new instance. I was able to successfully install and launch the app, even with browser windows open.

Apologies if I missed an instruction to ensure no browsers with integration plugins are running during installs. If I didn't, please consider adding an obvious note somewhere.

Thanks for all the help on this!

**Comment by droidmonkey** [comment]
Damn good catch

**Comment by tkoepsell** [comment]
Great!  Exiting all browsers BEFORE opening the dmg-file to install version 2.7.0 solved the problem I was having, too.

All is well again--thanks!

**Comment by phoerious** [comment]
> (Incidentally, the download page says that the current version for Intel processors is for MacOS 15+, which does not exist to my knowledge. The current MacOS version is 12.3.)

10.15, not 15.

I don't really understand why having keepassxc-proxy running would cause a codesigning issue to be raised upon opening KeePassXC. I mean, the upgrade changes the binary, obviously, but these are also two different processes.

**Comment by gxfgbo** [comment]
> Damn good catch

Thanks @droidmonkey ! I knew there was something I was missing. Just had to find it!



> I don't really understand why having keepassxc-proxy running would cause a codesigning issue to be raised upon opening KeePassXC. I mean, the upgrade changes the binary, obviously, but these are also two different processes.

@phoerious , I believe that the attempt to establish a communication channel between the app and the browser extension is causing grief for macOS Gatekeeper (perhaps the browser extension connection attempt is changing something in the app config before the app package install is accepted as valid by Gatekeeper? perhaps the attempted connection between the app and the extension itself? who knows...)

**Comment by phoerious** [comment]
The browser extension doesn't change anything. It just spins up a keepassxc-proxy process. As long as that is running, you may get "resource in use" errors when trying to replace the binary, but you shouldn't be seeing code signing issues.

**Comment by gxfgbo** [comment]
> The browser extension doesn't change anything. It just spins up a keepassxc-proxy process. As long as that is running, you may get "resource in use" errors when trying to replace the binary, but you shouldn't be seeing code signing issues.

Got it. I suspect then that Gatekeeper's policy prevents starting the proxy process before the DMG install is approved, and that Gatekeeper is flagging the attempted proxy start as a failed ID verification (which it isn't, of course).

In any case, it seems there's a way to work-around the issue without having to bypass Gatekeeper. Again, thanks for all the help.
